lidar_topic: /center/pandar
max_num_threads:  8
save_trajectory: true
suspend_on_failure: false
write_to_file: true

# ODOMETRY OPTIONS ----
odometry_options:
  voxel_size: 1.5 # The voxel size for the grid sampling of the new frame (before keypoints extraction)
  sample_voxel_size: 2 # The size of a voxel for the selection of `keypoints` by grid sampling
  max_distance: 200.0 # The threshold of the distance to suppress voxels from the map
  max_num_points_in_voxel: 20 # The maximum number of points per voxel of the map
  debug_print: false
  min_distance_points: 0.1
  distance_error_threshold: 5.0 # The motion of the sensor between two frames which is considered erroneous (stops the odometry)
  motion_compensation: 3 # The profile of the motion compensation (0=NONE, 1=CONSTANT_VELOCITY, 2=ITERATIVE, 3=CONTINUOUS)
  initialization: 1 #(0=NONE, 1=INIT_CONSTANT_VELOCITY)
  init_num_frames: 10
  robust_registration: false # Whether to use the robust layer (which will add computation time, trying again the registration in case of failures)
  robust_num_attempts: 3 # The maximum number of attempts performed in robust regime
  robust_full_voxel_threshold: 0.5
  robust_max_voxel_neighborhood: 4
  robust_threshold_relative_orientation: 2
  robust_threshold_ego_orientation: 2
  robust_minimal_level: 0
  log_to_file: false

  # CT_ICP OPTIONS ----
  ct_icp_options:
    size_voxel_map: 1.0 # The voxel size of in the voxel map
    num_iters_icp: 15 # The number of iterations of the ICP
    min_number_neighbors: 20 # The minimum number of neighbor points to define a valid neighborhood
    voxel_neighborhood: 1
    max_number_neighbors: 20
    max_dist_to_plane_ct_icp: 0.5
    debug_print: false
    point_to_plane_with_distortion: true
    distance: 1 # Options: [CT_POINT_TO_PLANE =1, POINT_TO_PLANE = 0]
    num_closest_neighbors: 1
    beta_location_consistency: 0.005
    beta_constant_velocity: 0.005
    beta_orientation_consistency: 0.02 # Trajectory Regularisation Weight to constrain the optimization
    beta_small_velocity: 0.00 # Trajectory Regularisation Weight to constrain the optimization

    solver: 1 # Options: [GN=Gauss-Newton = 0, CERES = 1]
    loss_function: 1 # Options: [0 = STANDARD, 1 = CAUCHY, 2 = HUBER, 3 = TOLERANT, 4 = TRUNCATED] (for CERES solver)
    ls_max_num_iters: 10 # The number of steps performed for each iteration of the ICP (for CERES solver)
    ls_num_threads:  8 # The number of threads to build and solve the least square system (for CERES solver)
    ls_sigma: 0.2 # The sigma parameter for loss CAUCHY, HUBER, TOLERANT, TRUNCATED (for CERES solver)
    ls_tolerant_min_threshold: 0.05 # The tolerant parameter for loss TOLERANT (for CERES solver)

    weight_alpha: 0.9
    weight_neighborhood: 0.1

    threshold_voxel_occupancy: 5 # The minimum number of points in a voxel to consider it in the neighbor search
    power_planarity: 2 # The power of planarity
    threshold_orientation_norm: 0.001 # Threshold on orientation changes (in degrees) for early termination of the ICP
    threshold_translation_norm: 0.001 # Threshold on distance changes (in m) for early termination of the ICP

    min_num_residuals: 200 # The minimum number of residuals for a valid ICP problem
    max_num_residuals: 1000 # The maximum number of residuals considered (if more keypoints exist, residuals are randomly sampled)
    weighting_scheme: 2
                                # PLANARITY = 0,      // Weighs residuals by their planarity coefficient
                                # NEIGHBORHOOD = 1,   // Weighs residuals by the distance to their neares neighbors
                                # ALL = 2             // Combines all weighting schemes with different coefficients
